File: packages/core/src/digest.ts
import { createHash } from 'node:crypto';
import { canonicalize } from './canonical.js';

export type DigestAlgorithm = 'sha-256';

const NODE_HASH_NAME: Record<DigestAlgorithm, string> = {
  'sha-256': 'sha256',
};

const DIGEST_PATTERN = /^(sha-256):([0-9a-f]{64})$/;

/** Computes an ACT digest string ("algorithm:hex") over raw bytes. */
export function digestBytes(
  bytes: Uint8Array | string,
  algorithm: DigestAlgorithm = 'sha-256',
): string {
  const hash = createHash(NODE_HASH_NAME[algorithm]).update(bytes).digest('hex');
  return `${algorithm}:${hash}`;
}

/** Computes an ACT digest over the RFC 8785 canonical bytes of a JSON value. */
export function digestCanonicalValue(
  value: unknown,
  algorithm: DigestAlgorithm = 'sha-256',
): string {
  return digestBytes(canonicalize(value), algorithm);
}

export interface ParsedDigest {
  algorithm: DigestAlgorithm;
  hex: string;
}

/** Parses and validates the "algorithm:hex" digest form. Throws on malformed or unregistered-algorithm input. */
export function parseDigest(digest: string): ParsedDigest {
  const match = DIGEST_PATTERN.exec(digest);
  if (!match) {
    throw new InvalidDigestError(digest);
  }
  return { algorithm: match[1] as DigestAlgorithm, hex: match[2] as string };
}

export function isValidDigestForm(digest: string): boolean {
  return DIGEST_PATTERN.test(digest);
}

/** Recomputes a digest over canonical bytes and compares to the claimed value in constant-ish time via string equality post-hash. */
export function verifyDigest(value: unknown, claimedDigest: string): boolean {
  const { algorithm } = parseDigest(claimedDigest);
  return digestCanonicalValue(value, algorithm) === claimedDigest;
}

export class InvalidDigestError extends Error {
  constructor(digest: string) {
    super(`Invalid or unregistered digest: ${JSON.stringify(digest)}`);
    this.name = 'InvalidDigestError';
  }
}

File: sdks/python/src/act_sdk/core/digest.py
"""Ports `packages/core/src/digest.ts`."""

from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from typing import Any

from .canonical import canonicalize

DigestAlgorithm = str  # only 'sha-256' is registered today

_DIGEST_PATTERN = re.compile(r"^(sha-256):([0-9a-f]{64})$")


def digest_bytes(data: bytes | str, algorithm: DigestAlgorithm = "sha-256") -> str:
    """Computes an ACT digest string ("algorithm:hex") over raw bytes."""
    if algorithm != "sha-256":
        raise InvalidDigestError(algorithm)
    raw = data.encode("utf-8") if isinstance(data, str) else data
    return f"sha-256:{hashlib.sha256(raw).hexdigest()}"


def digest_canonical_value(value: Any, algorithm: DigestAlgorithm = "sha-256") -> str:
    """Computes an ACT digest over the RFC 8785 canonical bytes of a JSON value."""
    return digest_bytes(canonicalize(value), algorithm)


@dataclass(frozen=True)
class ParsedDigest:
    algorithm: DigestAlgorithm
    hex: str


def parse_digest(digest: str) -> ParsedDigest:
    match = _DIGEST_PATTERN.match(digest)
    if not match:
        raise InvalidDigestError(digest)
    return ParsedDigest(algorithm=match.group(1), hex=match.group(2))


def is_valid_digest_form(digest: str) -> bool:
    return bool(_DIGEST_PATTERN.match(digest))


def verify_digest(value: Any, claimed_digest: str) -> bool:
    parsed = parse_digest(claimed_digest)
    return digest_canonical_value(value, parsed.algorithm) == claimed_digest


class InvalidDigestError(ValueError):
    def __init__(self, digest: str) -> None:
        super().__init__(f"Invalid or unregistered digest: {digest!r}")
